### DISCUSSION TITLE
Bitwise operator supported on Oracle?

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#11621] [Usage Questions]
Hello

Do you know if bit operators are working fine with Oracle?
I have encountered an error: `ORA-00920: invalid relational operator\nHelp: https://docs.oracle.com/error-help/db/ora-00920/"`
In the logs, I see that the operator has been replaced in the query by `&`, which is not supported by my database (`bitand` is the right syntax).

```python
# SQLAlchemy==2.0.28
# CREATE TABLE ... (
# "UNIX_RIGHTS" NUMBER(3,0) DEFAULT 1, 
# )
# unixrights= Column(SmallInteger, default=1)
unixrights= mapped_column('unixrights', SmallInteger, nullable=True, default=1)
filter = or_(User.unixrights.bitwise_and(1) != 0, User.unixrights.bitwise_and(2) != 0)
```

**Comment by zzzeek** [comment]
we dont have backend-specific tests for bitwise operators right now and these are hardcoded to emit the traditional `&|^` characters

**Reply by CaselIT** [reply]
created #11663

**Comment by CaselIT** [comment]
Oracle seems to be quite lacking regarding bitwise operators.
Version 19 does not have or and xor support, while 23 that I tried did.
And as far as I understand there is not bitwise_not or shift support in any version.

@nmoreaud do you know if they are supported and I've just missed them from the docs (I'm not an oracle expert)?

**Reply by nmoreaud** [reply]
> Oracle seems to be quite lacking regarding bitwise operators. Version 19 does not have or and xor support, while 23 that I tried did. And as far as I understand there is not bitwise_not or shift support in any version.
> 
> @nmoreaud do you know if they are supported and I've just missed them from the docs (I'm not an oracle expert)?

Sorry I have no idea, I don't know Oracle database subtleties and I don't have an Oracle support account.

**Reply by CaselIT** [reply]
ok, no worries, just thought to ask
